arangod/Aql/UpsertModifier.cpp

```diff
@@ -277,49 +277,133 @@ void UpsertModifier::accumulate(InputAqlItemRow& row) {
+      std::unique_lock<std::mutex> guard(self->_resultMutex);
+      try {
+        TRI_ASSERT(std::holds_alternative<Waiting>(self->_results));
+        if (std::holds_alternative<Waiting>(self->_results)) {
+          // get() will throw if opRes holds an exception, which is intended.
+          std::move(tryResult).get();
+          self->_results = HaveResult{};
+        } else {
+          // This can never happen.
+          using namespace std::string_literals;
+          auto state = std::visit(
+              overload{[&](NoResult) { return "NoResults"s; },
+                       [&](Waiting) { return "Waiting"s; },
+                       [&](HaveResult) { return "Result"s; },
+                       [&](std::exception_ptr const& ep) {
+                         auto what = std::string{};
+                         try {
+                           std::rethrow_exception(ep);
+                         } catch (std::exception const& ex) {
+                           what = ex.what();
+                         } catch (...) {
+                           // Exception unknown, give up immediately.
+                           LOG_TOPIC("adb0e", FATAL, Logger::AQL)
+                               << "Caught an exception while handling another "
+                                  "one, giving up.";
+                           FATAL_ERROR_ABORT();
+                         }
+                         return StringUtils::concatT("Exception: ", what);
+                       }},
+              self->_results);
+          auto message = StringUtils::concatT(
+              "Unexpected state when reporting modification result, expected "
+              "'Waiting' but got: ",
+              state);
+          LOG_TOPIC("3b0e1", ERR, Logger::AQL) << message;
+          if (std::holds_alternative<std::exception_ptr>(self->_results)) {
+            // Avoid overwriting an exception with another exception.
+            LOG_TOPIC("78c8b", FATAL, Logger::AQL)
+                << "Caught an exception while handling another one, giving up.";
+            FATAL_ERROR_ABORT();
+          }
+          THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL_AQL,
+                                         std::move(message));
+        }
+      } catch (...) {
+        auto exptr = std::current_exception();
+        self->_results = exptr;
+      }
+      return true;
+    });
+  });
+
+  return ExecutionState::WAITING;
+}
 
+futures::Future<futures::Unit> UpsertModifier::transactInternal(
+    transaction::Methods& trx) {
   auto toInsert = _insertAccumulator.closeAndGetContents();
   if (toInsert.isArray() && toInsert.length() > 0) {
-    _insertResults =
-        trx.insert(_infos._aqlCollection->name(), toInsert, _infos._options);
+    auto future = [&] {
+      auto guard = std::lock_guard(_trxMutex);
+      if (!_trxAlive) {
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_QUERY_KILLED);
+      }
+      return trx.insertAsync(_infos._aqlCollection->name(), toInsert,
```

Comment:
Is it intended that we hold `_trxMutex` here while executing the `insertAsync` call?
Same question for the following calls to `updateAsync` / `replaceAsync`.

Reply:
Yes, this is intentional. This serializes the calls to `insertAsync` (etc.) with `stopAndClear()`, which in turn makes sure that the transaction methods aren't destroyed during the call.
Note that `co_await`ing the Future `insertAsync` (etc.) returns is done after releasing the mutex, so it's not held during network requests. The transaction Methods themselves have to make sure that the actually asynchronous part of the code handles a concurrent destroy of the Methods properly (which my previous PRs should ensure).